apps/utils/cipher/cipher.main.kts

```diff
@@ -0,0 +1,505 @@
+        val keySuffix = secureText.utf8()
+        return "$keyPrefix:$keySuffix"
+    }
+}
+
+// Extension method that converts a ByteString into a BufferedSource.
+fun ByteString.toBuffer(): BufferedSource =
+    Buffer().readFrom(ByteArrayInputStream(toByteArray()))
+
+// Extension method that returns the next Block, if any, from a BufferedSource.
+fun BufferedSource.nextBlock(): Block? {
+    if (this.exhausted()) {
+        return null
+    }
+
+    var firstIndex = -1L
+    var firstToken = Token.None
+    for (token in Token.values()) {
+        if (token.bytes.size > 0) {
+            val index = this.indexOf(token.bytes)
+            if (index >= 0) {
+                if (firstIndex < 0 || index < firstIndex) {
+                    firstIndex = index
+                    firstToken = token
+//                println("$index $token")
+                }
+            }
+        }
+    }
+//    println("FOUND $firstIndex $firstToken")
+
+    val result = if (firstIndex < 0L) {
+        val bytes = this.readByteString()
+        Block(Token.None, bytes)
+    } else if (firstIndex == 0L) {
+        this.skip(firstToken.bytes.size.toLong())
+        Block(firstToken, EMPTY)
+    } else {
+        val bytes = this.readByteString(firstIndex)
+        this.skip(firstToken.bytes.size.toLong())
+        Block(firstToken, bytes)
+    }
+//    println(result)
+    return result
+}
+
+// Encapsulates KMS operations.
+class KmsCipher(private val endpoint: String?, region: String, keyArn: String) {
+    // AWS Crypto SDK client.
+    private val crypto = AwsCrypto.standard()
+
+    // Used to look up keys in KMS.
+    private val keyProvider = KmsMasterKeyProvider.builder()
+        .customRegionalClientSupplier(::createKmsClient)
+        .defaultRegion(Region.of(region))
+        .buildStrict(keyArn)
+
+    // Context info used when encrypting strings.
+    private val context = Collections.singletonMap("Application", APP_NAME)
+
+    // Creates a KMS SDK client object.
+    private fun createKmsClient(region: Region): KmsClient {
+        val builder = KmsClient.builder().region(region)
+        if (endpoint != null) {
+            builder.endpointOverride(URI.create(endpoint))
+        }
+        return builder.build()
+    }
+
+    // Encrypts the provided plain text and returns the base 64 encoded cipher text.
+    fun encrypt(plainText: ByteString): ByteString {
+        val encryptResult: CryptoResult<ByteArray, KmsMasterKey> =
+            crypto.encryptData(keyProvider, plainText.toByteArray(), context)
+        return Base64.getUrlEncoder().encodeToString(encryptResult.result).encodeUtf8()
+    }
+
+    // Decrypts the provided base 64 encoded cipher text and returns its equivalent plain text.
+    fun decrypt(cipherText: ByteString): ByteString {
+        val decryptResult = crypto.decryptData(keyProvider, Base64.getUrlDecoder().decode(cipherText.utf8()))
+        return decryptResult.result.toByteString()
+    }
```

Comment:
nit: `decrypt` appears to not work with KMS CMK aliases, instead requiring that the _true_ `key` (not `alias`) ARN is provided. Not a big deal, and definitely not something for this script to handle, but maybe something to explicitly call out in the help text?

Note that `encrypt` _does_ work with aliases. I thought this was surprising, but it might just be a quirk of KMS key aliases that I'm unfamiliar with.

Reply:
> nit: `decrypt` appears to not work with KMS CMK aliases, instead requiring that the _true_ `key` (not `alias`) ARN is provided. Not a big deal, and definitely not something for this script to handle, but maybe something to explicitly call out in the help text?

That's true! I've found working with the KMS API to be a little annoying at times due to the ambiguity surrounding where aliases and IDs can be used vs where Key ARNs are mandated–maybe it's not just me?

Here, we're only using a subset of the encryption-sdk's functionality. Even so, the developers made an intentional decision for us, and we aren't doing any kind of discovery, lookups, or mapping from less specific identifiers (i.e. aliases, key ids, alias arns) to their respective global identifiers of the key arn on decryption. [See this example for a tiny bit more discussion on this](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/java-example-code.html#java-example-strings).

From the help text:
```
[..snipped..]
        Options:
        --key ARN      : KMS key ARN (REQUIRED unless CIPHER_KEY is defined).
        --endpoint URI : Localhost endpoint URI for localstack KMS.
        --region name  : AWS region name.  Defaults to $DEFAULT_REGION.
[..snipped..] 
```

Maybe in a future iteration, we could consider checking input of the `--key` to ensure it's a valid KMS key arn before proceeding 🤔.

Reply:
Another feature to add next sprint!  Maybe we can do an explicit lookup of the ARN from an alias.  Failing that the validity check is an excellent idea.